pkg/controller/endpoints/endpoints_controller_test.go

```diff
@@ -0,0 +1,237 @@
+// is contributed by such licensors),
+//
+// without restriction, including without limitation the rights to copy, create
+// derivative works of, display, perform, and distribute the Software and make,
+// use, sell, offer for sale, import, export, have made, and have sold the
+// Software and the Larger Work(s), and to sublicense the foregoing rights on
+// either these or other terms.
+//
+// This license is subject to the following condition:
+// The above copyright notice and either this complete permission notice or at
+// a minimum a reference to the UPL must be included in all copies or
+// substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package endpoints_test
+
+import (
+	"context"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/operator-framework/operator-sdk/pkg/log/zap"
+	rhjmcv1beta1 "github.com/rh-jmc-team/container-jfr-operator/pkg/apis/rhjmc/v1beta1"
+	"github.com/rh-jmc-team/container-jfr-operator/pkg/controller/endpoints"
+	"github.com/rh-jmc-team/container-jfr-operator/test"
+)
+
+var _ = Describe("EndpointsController", func() {
+	var (
+		objs       []runtime.Object
+		client     client.Client
+		controller *endpoints.ReconcileEndpoints
+	)
+
+	JustBeforeEach(func() {
+		logf.SetLogger(zap.Logger())
+		s := test.NewTestScheme()
+
+		client = fake.NewFakeClientWithScheme(s, objs...)
+		controller = &endpoints.ReconcileEndpoints{
+			Client:     client,
+			Scheme:     s,
+			Reconciler: test.NewTestReconcilerNoServer(client),
+		}
+	})
+
+	AfterEach(func() {
+		objs = nil
+	})
+
+	Describe("reconciling a request", func() {
+		Context("successfully reconcile, without JMX auth", func() {
+			BeforeEach(func() {
+				objs = []runtime.Object{
+					test.NewContainerJFR(), test.NewContainerJFRService(),
+					test.NewTargetPod(), test.NewTestEndpoints(),
+				}
+			})
+			It("should create new flightrecorder", func() {
+				req := reconcile.Request{NamespacedName: types.NamespacedName{Name: "containerjfr", Namespace: "default"}}
```

Comment:
I think it makes sense to separate the Container JFR case from other Endpoints, because of the extra JMX auth steps involved and for general clarity.

For the typical case, we're reconciling Endpoints belonging to a user's application. Instead of using `test.NewContainerJFRService`, I would add a `test.NewTestService` that creates a Service called "test-svc". Likewise, rename the Endpoints created by `test.NewTestEndpoints` to "test-svc" (they have to be the same name).

Then for the Container JFR special case below (currently "with JMX auth"), use `test.NewContainerJFRService`, and create a `test.NewContainerJFREndpoints` and `test.NewContainerJFRPod` that create "containerjfr" and "containerjfr-pod" objects respectively.

I hope that all makes sense.